openc3/lib/openc3/microservices/trigger_group_microservice.rb

```diff
@@ -25,40 +25,47 @@
 require 'openc3/models/trigger_model'
 require 'openc3/topics/autonomic_topic'
 require 'openc3/utilities/authentication'
+require 'openc3/packets/json_packet'
 
 require 'openc3/script'
 
 module OpenC3
-
   class TriggerLoopError < TriggerError; end
 
   # Stored in the TriggerGroupShare this should be a thread safe
   # hash that triggers will be added, updated, and removed from
   class PacketBase
-
     def initialize(scope:)
       @scope = scope
       @mutex = Mutex.new
       @packets = Hash.new
     end
 
-    # ["#{@scope}__DECOM__{#{@target}}__#{@packet}"]
     def packet(target:, packet:)
       topic = "#{@scope}__DECOM__{#{target}}__#{packet}"
       @mutex.synchronize do
-        return Marshal.load( Marshal.dump(@packets[topic]) )
+        return nil unless @packets[topic]
+        # Deep copy the packet so it doesn't change under us
+        return Marshal.load( Marshal.dump(@packets[topic][-1]) )
       end
     end
 
-    def get(topic:)
+    def previous_packet(target:, packet:)
+      topic = "#{@scope}__DECOM__{#{target}}__#{packet}"
       @mutex.synchronize do
-        return Marshal.load( Marshal.dump(@packets[topic]) )
+        return nil unless @packets[topic] and @packets[topic].length == 2
+        # Deep copy the packet so it doesn't change under us
+        return Marshal.load( Marshal.dump(@packets[topic][0]) )
       end
     end
 
     def add(topic:, packet:)
       @mutex.synchronize do
-        @packets[topic] = packet
+        @packets[topic] ||= []
+        if @packets[topic].length == 2
+          @packets[topic].shift
+        end
+        @packets[topic].push(packet)
```

Comment:
I'm keeping a two packet history for the CHANGES and DOES NOT CHANGE comparisons